File: backend/src/domain/value-objects/PostalCode.ts
const SPANISH_POSTAL_CODE_PATTERN = /^(?:0[1-9]|[1-4]\d|5[0-2])\d{3}$/;

export class PostalCode {
	static readonly DEFAULT = PostalCode.create("35001");

	private constructor(public readonly value: string) {
		Object.freeze(this);
	}

	static create(value: string): PostalCode {
		if (!SPANISH_POSTAL_CODE_PATTERN.test(value)) {
			throw new Error(`Invalid Spanish postal code: "${value}"`);
		}

		return new PostalCode(value);
	}

	equals(other: PostalCode): boolean {
		return this.value === other.value;
	}

	toString(): string {
		return this.value;
	}
}

File: backend-dotnet/src/api/Domain/ValueObjects/PostalCode.cs
using System.Text.RegularExpressions;
using SharpMonads.Core;

namespace SupermarketPlanner.Api.Domain.ValueObjects;

public readonly record struct PostalCode
{
    private static readonly Regex SpanishPostalCodePattern =
        new(@"^(?:0[1-9]|[1-4]\d|5[0-2])\d{3}$", RegexOptions.Compiled);

    private readonly string value;

    private PostalCode(string value) => this.value = value;

    public static Either<DomainError, PostalCode> Create(string value)
    {
        if (!SpanishPostalCodePattern.IsMatch(value))
            return Either<DomainError, PostalCode>.FromLeft(
                DomainError.Validation(nameof(value), $"Invalid Spanish postal code: \"{value}\""));

        return Either<DomainError, PostalCode>.FromRight(new PostalCode(value));
    }

    public static PostalCode Default => new("35001");

    public string AsString() => value;
}
